.github/workflows/build-ubuntu16.04-HDFS.yml

```diff
@@ -82,6 +92,43 @@ jobs:
 
           source $GITHUB_WORKSPACE/scripts/ci/build_benchmarks.sh
 
+      - name: 'dump  core stacks'
+        if: ${{ always() == true   && startsWith(matrix.os, 'ubuntu-') == true }} # only run this job if the build step failed
+        run: |
+          # following contains answers with various possibilities for locating core files on various systems
+          # https://stackoverflow.com/questions/2065912/core-dumped-but-core-file-is-not-in-the-current-directory
+          #a 'gdb' is present there, not a file, other gdbs but seem assoc'd with non-general usage
+          #if [[ -f /usr/share/gdb ]]; then
+          #  dbgr="/usr/share/gdb"
+          #  dbgropts=" -q --core \$f --batch -ex \"thread apply all bt\" -ex \"quit\""
+          #fi
+          # lldb ref'd here located by doing trial in runner with find / -name "*lldb*" and sifting through the results.
+          if [[ -f /usr/lib/llvm-11/bin/lldb ]]; then
+            #lldb -c $f --batch -o 'bt all' -o 'quit'
+            dbgr="/usr/lib/llvm-11/bin/lldb"
+            #dbgropts=" -core \$f --batch -o 'bt all' -o 'quit'"
+          else
+            echo "debugger not found in previously seen location!"
```

Comment:
This is only checking for lldb, but ubuntu should have gdb right? Should we look for gdb in `/usr/bin`?

Reply:
Looked for one, there did not appear to be an available executable gdb that was not part of android or some other sub-something that could be used.  I didn't think it would be appropriate to try to use the android one.
There was, IIRC, a /usr/bin/gdb, -but-, it was a directory, not an executable...

Reply:
Suppose we could, at expense of overhead, try to install gdb.
But, the lldb, at least for now is there.  
The test, rather than just presumptive use, is because of possible changes in the installed software of the runner, there are I think clang10 and clang12 present, but neither of those installations seemed to have lldb installed in them, IIRC.
And the failure to find a gdb 'normal'ly confiured, sort of surprised me.

Reply:
The `/usr/bin/gdb` binary should always be part of the binary package `gdb` no matter which Ubuntu version you run. `lldb` is available to if you opt into clang's  `llvm`. Usually multiple compiler version are available; I currently gave `gdb` version 10 as `gcc` version is my default; in a Docker container running 16.04 (to match this setup) `gdb --version` says 7.11.1.

Reply:
FTR, trial installation of gdb required about 22secs, expect that could vary, but probably common (a Monday, 19:00-1930ish EST)
```
/usr/bin/time sudo apt install gdb
ii  libgdbm-compat4:amd64                                       1.18.1-5                                                            amd64        GNU dbm database routines (legacy support runtime version) 
ii  libgdbm6:amd64                                              1.18.1-5                                                            amd64        GNU dbm database routines (runtime version) 
ii  python3-gdbm:amd64                                          3.8.10-0ubuntu1~20.04                                               amd64        GNU dbm database support for Python 3.x

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  libbabeltrace1 libboost-regex1.71.0 libc6-dbg libipt2
  libsource-highlight-common libsource-highlight4v5 libxxhash0
Suggested packages:
  gdb-doc gdbserver
The following NEW packages will be installed:
  gdb libbabeltrace1 libboost-regex1.71.0 libc6-dbg libipt2
  libsource-highlight-common libsource-highlight4v5 libxxhash0
0 upgraded, 8 newly installed, 0 to remove and 13 not upgraded.
Need to get 14.7 MB of archives.
...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
5.99user 2.02system 0:22.45elapsed 35%CPU (0avgtext+0avgdata 128376maxresident)k
443112inputs+493072outputs (782major+374570minor)pagefaults 0swaps
++ which gdb
+ [[ -n /usr/bin/gdb ]]
++ which gdb
+ echo 'found gdb, /usr/bin/gdb'
found gdb, /usr/bin/gdb
```

Reply:
@Shelnutt2 there appears to be no gdb by default, can be installed, see above, took about 22secs in trial install.

Since there is one lldb available (and lldb is used on mac), do we want to install/use gdb, 
or stay with the lldb currently available?

Since there seem to be three installations of clang present (10/11/12), and the 'middle' one (11) has the lldb, I can speculate that lldb will likely remain until at least two major upgrades of clang ('til clang 14?) in the runner image.

Or, install gdb for use, and hope for indefinite stability regarding its availability?

Reply:
I do not think there is a `lldb`. I siggest to just ignore the Android SDK you are being handed for unclear reasons in the runner image you use.  Installing `gdb` is the right step, just how we install `cmake` or `git` or ... in other yaml files.  If you prefer you can install `lldb` but you then need to switch the compilation to using `clang` and `clang++`.

Reply:
@eddelbuettel FWIW, a functional lldb is there, at "/usr/lib/llvm-11/bin/lldb", 
and I can/am currently successfully using it by absolute path reference.

Reply:
Well this is Unix / Linux. Things are generally in the `$PATH`, if you find yourself unwrangling five levels this way something is possibly wrong.  It may that you have `clang-10` and/or `clang-11` installed and that the latter tucks this away.  

But I just did  `apt install lldb lldb-11` and now have 

```sh
root@f15c7180fa7a:/# which lldb-11
/usr/bin/lldb-11
root@f15c7180fa7a:/# lldb-11 --version
lldb version 11.0.0
root@f15c7180fa7a:/# 
root@f15c7180fa7a:/# which lldb   
/usr/bin/lldb
root@f15c7180fa7a:/# lldb --version
lldb version 10.0.0
root@f15c7180fa7a:/#
```

in a standard Ubuntu 20.04 container.  Easy choice between two different ones.

Reply:
@Shelnutt2 @eddelbuettel 
now installing gdb and using that to present stack backtraces, with fallback possibility to (currently) known lldb

Reply:
Nice.  I actually missed something at the start:  are you tickling core dumps on purpose here?  Do you also need to to enablem them -- as I recall some distros set the shell up via `ulimit -c 0` to not write any core files.

Reply:
Yes that is (now) being done.  
It was being done in a separate step, and I was getting core dumps in my earlier trials, but when I was verifying the gdb change, it was broken, a core dump was reported as generated but I could not find it.  
Apparently the ulimit -c unlimited needs to be done in the specific action one wants to see core dumps from (each separate bash invocation?),
and does not generally carry forward from a previous one.  So somewhere I damaged something in trying to clean up the core dump hacktivation code for production.

FWIW, I've (think) I've actually seen three different scenarios 1) an abort would be reported, no core dump reported [and no ulimit -c unlimited had been done anywhere], 2) an abort with core dump would be reported, but could not actually find the core file [and ulimit -c had been done in a previous step], and 3) [when ulimit -c unlimited done in the same step core dump expected] an abort with core dump was reported and I _could_ actually find the core file.

Reply:
I will have to check but I think distributions may just default (via global settings in `/etc/`) to have core files off my default.  That means that you have to either change it there (but can't for CI), or for the user resource files (usually files starting with a dot below `$HOME` aka `~`) or if you can't then for each separate shell invocation.  Which gets us to the CI runs.  There may be global options, else you may have to indeed to it for each shell that is spawned (and I would guess different 'jobs' or 'steps' in one YAML might inherit from a global one -- if in doubt set it "everywhere").  You can get the "everywhere" behavior by something like `echo "ulimit -c -1" >> ~/.profile` (or maybe into `~/.bashrc` or ... -- there are sadly different resource files for different types of shell invocation, login vs non-login etc).